### ISSUE TITLE
Proxy between Network Namespaces

### ISSUE THREAD

**Issue by VasiliPupkin256** [z3APA3A/3proxy#486]
Please consider implement proxying between network namespaces. Despite broad info that process can only have sockets in one network namespace actually it can have opened sockets in multiple namespaces. All the opened sockets aren't affected by a `setns` system call. A socket is a network device with associated network namespace and new accepted connections are also belongs to the original socket namespace.

So the procedure is:

- Create a listen proxy socket

- Call `setns()` function and move the process into a new namespace

- Do chroot and other privilege deescalation stuff

- Do the main proxy loop as usual, accept connections on original socket and make new connections in the new network namespace

I don't know how to arrange this with a complex configs of multiple proxy interfaces but it is possible to make a simple proxy between namespaces this way without the performance overhead and setup pain associated with virtual `veth` interfaces between namespaces.

**Comment by z3APA3A** [comment]
Is it enough to implement a

setns path <flags>

config command? (3proxy guarantees the order of config commands execution)

**Comment by z3APA3A** [comment]
Or clone() support is also required? Can you describe a complete usecase with namespace creation?

**Comment by VasiliPupkin256** [comment]
Maybe a new `setns path` command would be enough... Does 3proxy reopen listen connection? What about outgoing DNS requests?

My usecase is proxying to a vpn interface that is opened inside a network namespace, see https://www.wireguard.com/netns/ 

The other possible usecase is to proxy between a host and a virtual machine that is running inside a container.

**Comment by VasiliPupkin256** [comment]
Does it verify the existence of external address and interface (options `-e` and `-De`) when it starts a proxy but no connections were made yet? This address and interface may not be reachable in original namespace before the `setns` call.

**Comment by z3APA3A** [comment]
3proxy reopens listening sockets only on configuration reload (in the case of the failure it performs repeating attempts to listen()). External address is not used before client request is received. See
https://3proxy.ru/faqr.asp#INTEXT
https://3proxy.ru/faqe.asp#INTEXT

**Comment by z3APA3A** [comment]
> Create a listen proxy socket
> 
> Call setns() function and move the process into a new namespace
> 
> Do chroot and other privilege deescalation stuff
> 


nope. I think this approach does not work, because setns() sets a namespace for the current thread and calling it on config affects a main thread, while proxying between namespaces probably requires ns to be set in a proxy thread. Probably, for this to work setting namespace should be an option for the service (like 'socks' or 'proxy') rather than config option to achieve this task.

also, currently I don't understand, if it can be usefull for a proxy thread to have ability to set a ns before opening the socket and after opening socket, or first case is never required.

I'm not sure with chroot() either and if chroot works with threads in different namespaces.

**Comment by VasiliPupkin256** [comment]
Oh.. threads. Yes it should be called inside the thread then.

>  if it can be usefull for a proxy thread to have ability to set a ns before opening the socket 

The original proxy namespace can always be set running 3proxy with `ip netns exec ...` command for example. but it can be implemented though the config for simplicity, `setns` can be called multiple times. Why not

I am also thinking about socks UDP ASSOCIATE request type, it can be broken by this. If I recall correctly this request type requires server to open a new UDP socket on a listening interface.

**Comment by phantomcraft** [comment]
HAProxy already implemented support for network namespaces: https://fossies.org/linux/haproxy/doc/network-namespaces.txt

But it requires root privileges.

**Comment by z3APA3A** [comment]
https://github.com/3proxy/3proxy/commit/51a82613a6758450ca9823d8846d42570e6b5b85

**Comment by alpominth** [comment]
@z3APA3A

Wonderfull feature, but not that is very invasive by requiring CAP_SYS_ADMIN, IMO the best way  to avoid requiring root or elevated privileges way is to write a small cross namespace socket library and a daemon that spawns a Unix socket, so that the socket is invoked by the library and passed via Unix socket to the daemon that creates the socket inside the target netns, it could work perfectly with incoming and outgoing connections. Note that the daemon requires root only for the command <ip netns exec ...> or <nsenter -t (PID) -n>, by running with sudo, it can run as a unprivileged user.

FYI, this daemon and cross namespace socket library already exist: https://github.com/koro666/xsocket

Its use is ridiculously simple to implement, just replace socket() syscall with: `int xsocket(const char* path, int domain, int type, int, protocol);`

As @koro666 said, the library (xsocket) is public domain as long as you don't fork for creating a similar commercial software, you could just embed the xsocket into 3proxy repository and use it as part of 3proxy.

Changing the subject, you missed one point in this implementation, just switching 3proxy to a netns, creating the socket inside it and binding and bringing back to original netns is not enough to make SOCKS5 work (or maybe another protocols) properly, because of UDP ASSOCIATE: everytime 3proxy receives a TCP message control requesting UDP ASSOCIATE, 3proxy should pause their threads and switch to the netns AGAIN to create the UDP Listening point inside the target netns, it's because the UDP Associate ports are dynamically assigned (unless you implement a way to make 3proxy listen on a fixed and persistent UDP port).

See this: https://github.com/folbricht/routedns/pull/499

`"Uses runtime.LockOSThread() + unix.Setns() + CLONE_NEWNET; sockets retain namespace affinity after thread switches back"`

Threads should be locked before switching 3proxy to a new netns, if one wants to use UDP with SOCKS5 3proxy, 3proxy will use this assumption everytime.

**Comment by z3APA3A** [comment]
3proxy supports unix sockets directly as both listening socket and as tcppm/parent destination, so it can be done without 3rd party library. UDP ASSOCIATE does not work with unix sockets either, but I think it can easiliy be fixed, if required.

**Comment by alpominth** [comment]
@z3APA3A 

>3proxy supports unix sockets directly as both listening socket and as tcppm/parent destination, so it can be done without 3rd party library.

I wouldn't run 3proxy as root even if netns option needs this, I saw you saying the same, that 3proxy shouldn't be run as root.

A new binary like "`3proxy_netnssrv`" that handles the socket creation inside the netns without requiring elevated privileges (except by the <ip netns exec somenetns sudo -u someuser ... > invocation) would be a good idea, but it would be a good idea also that the Unix socket of "`3proxy_netnssrv`" should be of type SOCK_SEQPACKET as it delivers the messages in the order sent, it's just some parameters like this: https://github.com/koro666/xsocket/blob/master/protocol.h

_When xsocket is called to create a socket, it connects to the xsocket-server instance through its UNIX socket and sends a message containing the requested domain, type and protocol.
Upon reception of such a message, the server creates the requested socket, then passes it back using SCM_RIGHTS to the client. In case of failure, an error code is returned instead._

>UDP ASSOCIATE does not work with unix sockets either, but I think it can easiliy be fixed, if required.

There is two options of using the SOCK_DGRAM type, one is creating a new thread per "connection" (it's lame of course), and the another is creating multiple file descriptors (additional Unix sockets) for each "connection", but a special care should be taken to delete all the temporary child unix sockets after 1 or 2 minutes of timeout.

```
root@localhost:/home/user# /dev/shm/3proxy/bin/3proxy_socks -l -ni/var/run/netns/nsx
260504191101.087 1080 00000 - 0.0.0.0:0 0.0.0.0:0 0 0 0 Accepting connections [9603/139742126651200]
260504191103.719 1080 00000 - 127.0.0.1:54296 127.0.0.1:40444 0 0 0 UDPMAP 127.0.0.1:40444
260504191103.722 1080 00000 - 127.0.0.1:54298 127.0.0.1:32835 0 0 0 UDPMAP 127.0.0.1:32835
260504191103.722 1080 00000 - 127.0.0.1:54310 127.0.0.1:38434 0 0 0 UDPMAP 127.0.0.1:38434
```

```
root@localhost:/home/user# netstat -nlp 
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 0.0.0.0:1080            0.0.0.0:*               LISTEN      9603/3proxy_socks   
Active UNIX domain sockets (only servers)
Proto RefCnt Flags       Type       State         I-Node   PID/Program name     Path
root@localhost:/home/user# SOCKS_SERVER="127.0.0.1:1080" socksify dig @1.1.1.1
;; communications error to 1.1.1.1#53: connection refused
;; communications error to 1.1.1.1#53: connection refused
;; communications error to 1.1.1.1#53: connection refused

; <<>> DiG 9.20.22-1-Debian <<>> @1.1.1.1
; (1 server found)
;; global options: +cmd
;; no servers could be reached
```

Just switching the 3proxy thread to a new netns, binding the socket there and returning back to original netns is not enough to make UDP Associate work with 3proxy inside a netns, this process should be repeated everytime a new temporary UDP listening point is spawned by the program, of course by freezing the threads before and releasing afterwards.
